# Text Classifier

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import keras
from keras.datasets import reuters

Using TensorFlow backend.


In [3]:
(x_train, y_train) , (x_test, y_test) = reuters.load_data(num_words= None, test_split=0.2 )

In [4]:
word_index= reuters.get_word_index()

In [5]:
y_train.shape

(8982,)

In [6]:
x_test.shape

(2246,)

In [7]:
num_classes= max(y_train)+1
print("Number of classes {}".format(num_classes))

Number of classes 46


In [8]:
print(x_train[0])

[1, 27595, 28842, 8, 43, 10, 447, 5, 25, 207, 270, 5, 3095, 111, 16, 369, 186, 90, 67, 7, 89, 5, 19, 102, 6, 19, 124, 15, 90, 67, 84, 22, 482, 26, 7, 48, 4, 49, 8, 864, 39, 209, 154, 6, 151, 6, 83, 11, 15, 22, 155, 11, 15, 7, 48, 9, 4579, 1005, 504, 6, 258, 6, 272, 11, 15, 22, 134, 44, 11, 15, 16, 8, 197, 1245, 90, 67, 52, 29, 209, 30, 32, 132, 6, 109, 15, 17, 12]


**Shows most frequent words**

In [9]:
print(y_train[0])

3


In [10]:
word_index['happy']

4041

In [11]:
index_to_word={}
for key, value in word_index.items():
    index_to_word[value] = key

In [12]:
index_to_word[4041]

'happy'

In [13]:
print(' ' . join([index_to_word[x] for x in x_train[0]]))

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs


In [14]:
print(' ' . join([index_to_word[x] for x in x_train[5]]))

the in has would seven time 35 we're but 18 national adjusted with growth received vs been reuter growth received vs been have five 1 35 exchange farm same with did dlr vs been reuter did dlr vs have five so but 18 premiums program their cts voting with 23 standard vs been reuter 23 standard vs been have five their a exchange farm cts 35 with did offering vs been reuter did offering vs have five so national 13 a but 18 or levels with shareholders raised vs been reuter sources raised vs been have five 13 a exchange farm or 35 with any businesses vs been reuter any addition vs have five pct dlrs


## Model

In [15]:
from keras.preprocessing.text import Tokenizer

max_words = 10000
tokenizer= Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode= 'binary')
x_test = tokenizer.sequences_to_matrix(x_test, mode= 'binary')

y_train = keras.utils.to_categorical (y_train, num_classes)
y_test = keras.utils.to_categorical (y_test, num_classes)

In [16]:
x_train.shape

(8982, 10000)

In [17]:
print(x_train[0])

[0. 1. 0. ... 0. 0. 0.]


In [18]:
y_train.shape

(8982, 46)

In [19]:
print(y_train[0])

[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [20]:
from keras.models import Sequential
from keras.layers import Dense,Dropout, Activation

model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [21]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.metrics_names)

['loss', 'acc']


In [22]:
batch_size = 64
epochs = 5

history = model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, verbose=1, validation_split=0.2)
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 7185 samples, validate on 1797 samples
Epoch 1/5
7185/7185 [==============================] - 11s 2ms/step - loss: 1.5151 - acc: 0.6793 - val_loss: 0.9768 - val_acc: 0.7913
Epoch 2/5
7185/7185 [==============================] - 10s 1ms/step - loss: 0.6156 - acc: 0.8676 - val_loss: 0.8278 - val_acc: 0.8169
Epoch 3/5
7185/7185 [==============================] - 10s 1ms/step - loss: 0.3257 - acc: 0.9294 - val_loss: 0.8122 - val_acc: 0.8203
Epoch 4/5
7185/7185 [==============================] - 10s 1ms/step - loss: 0.2240 - acc: 0.9489 - val_loss: 0.8484 - val_acc: 0.8197
Epoch 5/5
2246/2246 [==============================] - 1s 225us/step
Test loss: 0.9022383940400338
Test accuracy: 0.8058771148708815


**Training accuracy  starts from 0.67 to 0.95<br>
Validation accuracy is 0.81 <br>
Test accuracy is 0.80 **

In [24]:
len (x_test) * 0.805877

1809.999742

**We trained a model on 2246 test samples. It has classified about 80.5 % correctly.<br>
So we have classified 1809 of them correctly**